'use strict';


var pmid_items = module.exports.pmid_items = [
    {
        "id": 23294250,
        "doi": "10.1111/tbj.12077",
        "authors": [
            "Beggs AD",
            "Kousparos G",
            "Hodgson SV"
        ],
        "title": "Loss of mismatch repair protein expression in breast carcinoma in patients with Lynch Syndrome: report of two cases",
        "abstract": "Lynch Syndrome, an autosomal dominantly inherited cancer predisposition syndrome is not typically associated with development of breast cancer. We present two cases of loss of mismatch protein expression in breast cancer cases in patients with Lynch Syndrome and discuss the literature surrounding the subject.",
        "nlm_title": "The Breast Journal",
        "specifier": "19(2):193-5"
    },
    {
        "id": 22034109,
        "doi": "10.1002/gcc.20935",
        "authors": [
            "Buerki N",
            "Gautier L",
            "Kovac M",
            "Marra G",
            "Buser M",
            "Mueller H",
            "Heinimann K"
        ],
        "title": "Evidence for breast cancer as an integral part of Lynch syndrome",
        "abstract": "Lynch syndrome, an autosomal dominant cancer predisposition caused by mutations in DNA mismatch repair (MMR) genes, mainly mainly mutL homolog 1, OMIM 120436 (MLH1) and mutS homolog 2, OMIM 609309 (MSH2), encompasses a tumor spectrum including primarily gastrointestinal, endometrial, and ovarian cancer. This study aimed at clarifying the heavily debated issue of breast cancer being part of Lynch syndrome. Detailed clinical data on cancer occurrence in Swiss female MLH1/MSH2 mutation carriers were gathered, all available breast cancer specimens assessed for molecular evidence for MMR deficiency (i.e., microsatellite instability (MSI), MMR protein expression, and somatic (epi)genetic MMR gene alterations) and compiled with the scarce molecular data available from the literature. Seventy unrelated Swiss Lynch syndrome families were investigated comprising 632 female family members at risk of which 92 were genetically verified mutation carriers (52 MLH1 and 40 MSH2). On contrast to endometrial and ovarian cancer, which occurred significantly more often and at younger age in MLH1/MSH2 mutation carriers (median 50.5 and 49.0 years; P < 0.00001), overall cumulative breast cancer incidence closely mirrored the one in the Swiss population (56.5 years). Six (85.7%) of seven breast cancer specimens available for molecular investigations displayed the hallmarks of MMR deficiency. Combined with data from the literature, MSI was present in 26 (70.3%) of 37 and altered MMR protein expression in 16 (72.7%) of 22 breast cancer specimens from MLH1/MSH2 mutation carriers. These findings, thus, provide strong molecular evidence for a pivotal role of MMR deficiency in breast cancer development in Lynch syndrome.",
        "nlm_title": "Genes Chromosomes Cancer",
        "specifier": "51(1):83-91"
    },
    {
        "id": 23666231,
        "doi": "10.1007/s10689-013-9651-x",
        "authors": [
            "Casey MJ",
            "Bewtra C",
            "Lynch HT",
            "Snyder C",
            "Stacy M",
            "Watson P"
        ],
        "title": "Phenotypic heterogeneity of hereditary gynecologic cancers: a report from the Creighton hereditary cancer registry",
        "abstract": "To determine the validity of observations suggesting a significant dichotomy of gynecologic cancers determined by linkage to specific genetic defects associated with two major autosomal dominant hereditary cancer syndromes; the Creighton University Hereditary Cancer Registry was searched for female carriers of germ line mutations in BRCA1 and BRCA2, associated with the Hereditary Breast Ovarian Cancer syndrome, and in the mismatch repair (MMR) genes MLH1, MSH2 and MSH6, associated with Lynch syndrome, who were registered with invasive uterine, ovarian, fallopian tube or peritoneal cancers between January 1, 1959 and December 31, 2010. From 217 such cases, a total of 174 subjects, consisting of 95 BRCA1 and BRCA2 mutation carriers and 79 carriers of mutations in MMR genes, were identified who had current signed Health Insurance Portability and Accountability Act forms and complete primary diagnostic pathology reports and clinical records. Data meticulously extracted from these cases were categorized and statistically analyzed. There were highly significant differences between carriers of BRCA1 and BRCA2 mutations and carriers of MMR gene mutations in the proportion of serous carcinomas compared with endometrioid carcinomas of the uterus, including cervix and endometium (p < 0.002), ovaries (p < 0.001) and overall, including fallopian tube and peritoneum cancers (p < 0.001). Endometrioid carcinoma was found in one and transitional carcinoma in another of the 14 BRCA1 mutation carriers with fallopian tube cancer, and endometrioid carcinoma was found in two of four MMR gene mutation carriers with fallopian tube cancers. All other fallopian tube cancers were serous carcinomas. Seven BRCA1 and one BRCA2 mutation carriers were diagnosed with primary peritoneal serous carcinoma; no peritoneal carcinomas were registered in MMR gene mutation carriers. Nine of 14 gynecologic cancers with associated endometriosis in mutation carriers were endometrioid or endometrioid mixed carcinomas compared with just three of other histologic types. Primary breast cancers, that characterize the HBOC syndrome, were much more frequent in BRCA1 and BRCA2 mutation carriers; while multiple gynecologic cancers and associated colorectal and urinary tract cancers, which are features of Lynch syndrome, were more common in MMR gene mutation carriers. Both serous and endometrioid carcinomas were diagnosed in MMR gene mutation carriers at significantly younger ages than in BRCA1 and BRCA2 mutation carriers (p < 0.0006). These findings confirm a clear dichotomy of uterine, ovarian and fallopian tube cancers associated with inheritance of mutations in BRCA1 and BRCA2 contrasted with inheritance of MMR gene mutations. This opens possibilities for new approaches to molecular genetic research into carcinogenic pathways and raises important new considerations regarding counseling, screening, prophylaxis and treatment of mutation carriers.",
        "nlm_title": "Familial Cancer",
        "specifier": "12(4):719-40"
    },
    {
        "id": 24549055,
        "doi": "10.1038/ejhg.2014.16",
        "authors": [
            "Castéra L",
            "Krieger S",
            "Rousselin A",
            "Legros A",
            "Baumann JJ",
            "Bruet O",
            "Brault B",
            "Fouillet R",
            "Goardon N",
            "Letac O",
            "Baert-Desurmont S",
            "Tinat J",
            "Bera O",
            "Dugast C",
            "Berthet P",
            "Polycarpe F",
            "Layet V",
            "Hardouin A",
            "Frébourg T",
            "Vaur D"
        ],
        "title": "Next-generation sequencing for the diagnosis of hereditary breast and ovarian cancer using genomic capture targeting multiple candidate genes",
        "abstract": "To optimize the molecular diagnosis of hereditary breast and ovarian cancer (HBOC), we developed a next-generation sequencing (NGS)-based screening based on the capture of a panel of genes involved, or suspected to be involved in HBOC, on pooling of indexed DNA and on paired-end sequencing in an Illumina GAIIx platform, followed by confirmation by Sanger sequencing or MLPA/QMPSF. The bioinformatic pipeline included CASAVA, NextGENe, CNVseq and Alamut-HT. We validated this procedure by the analysis of 59 patients' DNAs harbouring SNVs, indels or large genomic rearrangements of BRCA1 or BRCA2. We also conducted a blind study in 168 patients comparing NGS versus Sanger sequencing or MLPA analyses of BRCA1 and BRCA2. All mutations detected by conventional procedures were detected by NGS. We then screened, using three different versions of the capture set, a large series of 708 consecutive patients. We detected in these patients 69 germline deleterious alterations within BRCA1 and BRCA2, and 4 TP53 mutations in 468 patients also tested for this gene. We also found 36 variations inducing either a premature codon stop or a splicing defect among other genes: 5/708 in CHEK2, 3/708 in RAD51C, 1/708 in RAD50, 7/708 in PALB2, 3/708 in MRE11A, 5/708 in ATM, 3/708 in NBS1, 1/708 in CDH1, 3/468 in MSH2, 2/468 in PMS2, 1/708 in BARD1, 1/468 in PMS1 and 1/468 in MLH3. These results demonstrate the efficiency of NGS in performing molecular diagnosis of HBOC. Detection of mutations within other genes than BRCA1 and BRCA2 highlights the genetic heterogeneity of HBOC.",
        "nlm_title": "European journal of human genetics : EJHG",
        "specifier": "22(11):1305-13"
    },
    {
        "id": 21356067,
        "doi": "10.1186/bcr2832",
        "authors": [
            "Kuusisto KM",
            "Bebel A",
            "Vihinen M",
            "Schleutker J",
            "Sallinen SL"
        ],
        "title": "Screening for BRCA1, BRCA2, CHEK2, PALB2, BRIP1, RAD50, and CDH1 mutations in high-risk Finnish BRCA1/2-founder mutation-negative breast and/or ovarian cancer individuals",
        "abstract": "INTRODUCTION: Two major high-penetrance breast cancer genes, BRCA1 and BRCA2, are responsible for approximately 20% of hereditary breast cancer (HBC) cases in Finland. Additionally, rare mutations in several other genes that interact with BRCA1 and BRCA2 increase the risk of HBC. Still, a majority of HBC cases remain unexplained which is challenging for genetic counseling. We aimed to analyze additional mutations in HBC-associated genes and to define the sensitivity of our current BRCA1/2 mutation analysis protocol used in genetic counseling. METHODS: Eighty-two well-characterized, high-risk hereditary breast and/or ovarian cancer (HBOC) BRCA1/2-founder mutation-negative Finnish individuals, were screened for germline alterations in seven breast cancer susceptibility genes, BRCA1, BRCA2, CHEK2, PALB2, BRIP1, RAD50, and CDH1. BRCA1/2 were analyzed by multiplex ligation-dependent probe amplification (MLPA) and direct sequencing. CHEK2 was analyzed by the high resolution melt (HRM) method and PALB2, RAD50, BRIP1 and CDH1 were analyzed by direct sequencing. Carrier frequencies between 82 (HBOC) BRCA1/2-founder mutation-negative Finnish individuals and 384 healthy Finnish population controls were compared by using Fisher's exact test. In silico prediction for novel missense variants effects was carried out by using Pathogenic-Or-Not -Pipeline (PON-P). RESULTS: Three previously reported breast cancer-associated variants, BRCA1 c.5095C > T, CHEK2 c.470T > C, and CHEK2 c.1100delC, were observed in eleven (13.4%) individuals. Ten of these individuals (12.2%) had CHEK2 variants, c.470T > C and/or c.1100delC. Fourteen novel sequence alterations and nine individuals with more than one non-synonymous variant were identified. One of the novel variants, BRCA2 c.72A > T (Leu24Phe) was predicted to be likely pathogenic in silico. No large genomic rearrangements were detected in BRCA1/2 by multiplex ligation-dependent probe amplification (MLPA). CONCLUSIONS: In this study, mutations in previously known breast cancer susceptibility genes can explain 13.4% of the analyzed high-risk BRCA1/2-negative HBOC individuals. CHEK2 mutations, c.470T > C and c.1100delC, make a considerable contribution (12.2%) to these high-risk individuals but further segregation analysis is needed to evaluate the clinical significance of these mutations before applying them in clinical use. Additionally, we identified novel variants that warrant additional studies. Our current genetic testing protocol for 28 Finnish BRCA1/2-founder mutations and protein truncation test (PTT) of the largest exons is sensitive enough for clinical use as a primary screening tool.",
        "nlm_title": "Breast cancer research : BCR",
        "specifier": "13(1):R20"
    }
];


var country_codes = module.exports.country_codes = [
    {code: "AF", name: "Afghanistan"},
    {code: "AX", name: "Åland Islands"},
    {code: "AL", name: "Albania"},
    {code: "DZ", name: "Algeria"},
    {code: "AS", name: "American Samoa"},
    {code: "AD", name: "Andorra"},
    {code: "AO", name: "Angola"},
    {code: "AI", name: "Anguilla"},
    {code: "AQ", name: "Antarctica"},
    {code: "AG", name: "Antigua and Barbuda"},
    {code: "AR", name: "Argentina"},
    {code: "AM", name: "Armenia"},
    {code: "AW", name: "Aruba"},
    {code: "AU", name: "Australia"},
    {code: "AT", name: "Austria"},
    {code: "AZ", name: "Azerbaijan"},
    {code: "BS", name: "Bahamas"},
    {code: "BH", name: "Bahrain"},
    {code: "BD", name: "Bangladesh"},
    {code: "BB", name: "Barbados"},
    {code: "BY", name: "Belarus"},
    {code: "BE", name: "Belgium"},
    {code: "BZ", name: "Belize"},
    {code: "BJ", name: "Benin"},
    {code: "BM", name: "Bermuda"},
    {code: "BT", name: "Bhutan"},
    {code: "BO", name: "Bolivia, Plurinational State of"},
    {code: "BQ", name: "Bonaire, Sint Eustatius and Saba"},
    {code: "BA", name: "Bosnia and Herzegovina"},
    {code: "BW", name: "Botswana"},
    {code: "BV", name: "Bouvet Island"},
    {code: "BR", name: "Brazil"},
    {code: "IO", name: "British Indian Ocean Territory"},
    {code: "BN", name: "Brunei Darussalam"},
    {code: "BG", name: "Bulgaria"},
    {code: "BF", name: "Burkina Faso"},
    {code: "BI", name: "Burundi"},
    {code: "KH", name: "Cambodia"},
    {code: "CM", name: "Cameroon"},
    {code: "CA", name: "Canada"},
    {code: "CV", name: "Cape Verde"},
    {code: "KY", name: "Cayman Islands"},
    {code: "CF", name: "Central African Republic"},
    {code: "TD", name: "Chad"},
    {code: "CL", name: "Chile"},
    {code: "CN", name: "China"},
    {code: "CX", name: "Christmas Island"},
    {code: "CC", name: "Cocos (Keeling) Islands"},
    {code: "CO", name: "Colombia"},
    {code: "KM", name: "Comoros"},
    {code: "CG", name: "Congo"},
    {code: "CD", name: "Congo, the Democratic Republic of the"},
    {code: "CK", name: "Cook Islands"},
    {code: "CR", name: "Costa Rica"},
    {code: "CI", name: "Côte d'Ivoire"},
    {code: "HR", name: "Croatia"},
    {code: "CU", name: "Cuba"},
    {code: "CW", name: "Curaçao"},
    {code: "CY", name: "Cyprus"},
    {code: "CZ", name: "Czech Republic"},
    {code: "DK", name: "Denmark"},
    {code: "DJ", name: "Djibouti"},
    {code: "DM", name: "Dominica"},
    {code: "DO", name: "Dominican Republic"},
    {code: "EC", name: "Ecuador"},
    {code: "EG", name: "Egypt"},
    {code: "SV", name: "El Salvador"},
    {code: "GQ", name: "Equatorial Guinea"},
    {code: "ER", name: "Eritrea"},
    {code: "EE", name: "Estonia"},
    {code: "ET", name: "Ethiopia"},
    {code: "FK", name: "Falkland Islands (Malvinas)"},
    {code: "FO", name: "Faroe Islands"},
    {code: "FJ", name: "Fiji"},
    {code: "FI", name: "Finland"},
    {code: "FR", name: "France"},
    {code: "GF", name: "French Guiana"},
    {code: "PF", name: "French Polynesia"},
    {code: "TF", name: "French Southern Territories"},
    {code: "GA", name: "Gabon"},
    {code: "GM", name: "Gambia"},
    {code: "GE", name: "Georgia"},
    {code: "DE", name: "Germany"},
    {code: "GH", name: "Ghana"},
    {code: "GI", name: "Gibraltar"},
    {code: "GR", name: "Greece"},
    {code: "GL", name: "Greenland"},
    {code: "GD", name: "Grenada"},
    {code: "GP", name: "Guadeloupe"},
    {code: "GU", name: "Guam"},
    {code: "GT", name: "Guatemala"},
    {code: "GG", name: "Guernsey"},
    {code: "GN", name: "Guinea"},
    {code: "GW", name: "Guinea-Bissau"},
    {code: "GY", name: "Guyana"},
    {code: "HT", name: "Haiti"},
    {code: "HM", name: "Heard Island and McDonald Islands"},
    {code: "VA", name: "Holy See (Vatican City State)"},
    {code: "HN", name: "Honduras"},
    {code: "HK", name: "Hong Kong"},
    {code: "HU", name: "Hungary"},
    {code: "IS", name: "Iceland"},
    {code: "IN", name: "India"},
    {code: "ID", name: "Indonesia"},
    {code: "IR", name: "Iran, Islamic Republic of"},
    {code: "IQ", name: "Iraq"},
    {code: "IE", name: "Ireland"},
    {code: "IM", name: "Isle of Man"},
    {code: "IL", name: "Israel"},
    {code: "IT", name: "Italy"},
    {code: "JM", name: "Jamaica"},
    {code: "JP", name: "Japan"},
    {code: "JE", name: "Jersey"},
    {code: "JO", name: "Jordan"},
    {code: "KZ", name: "Kazakhstan"},
    {code: "KE", name: "Kenya"},
    {code: "KI", name: "Kiribati"},
    {code: "KP", name: "Korea, Democratic People's Republic of"},
    {code: "KR", name: "Korea, Republic of"},
    {code: "KW", name: "Kuwait"},
    {code: "KG", name: "Kyrgyzstan"},
    {code: "LA", name: "Lao People's Democratic Republic"},
    {code: "LV", name: "Latvia"},
    {code: "LB", name: "Lebanon"},
    {code: "LS", name: "Lesotho"},
    {code: "LR", name: "Liberia"},
    {code: "LY", name: "Libya"},
    {code: "LI", name: "Liechtenstein"},
    {code: "LT", name: "Lithuania"},
    {code: "LU", name: "Luxembourg"},
    {code: "MO", name: "Macao"},
    {code: "MK", name: "Macedonia, the former Yugoslav Republic of"},
    {code: "MG", name: "Madagascar"},
    {code: "MW", name: "Malawi"},
    {code: "MY", name: "Malaysia"},
    {code: "MV", name: "Maldives"},
    {code: "ML", name: "Mali"},
    {code: "MT", name: "Malta"},
    {code: "MH", name: "Marshall Islands"},
    {code: "MQ", name: "Martinique"},
    {code: "MR", name: "Mauritania"},
    {code: "MU", name: "Mauritius"},
    {code: "YT", name: "Mayotte"},
    {code: "MX", name: "Mexico"},
    {code: "FM", name: "Micronesia, Federated States of"},
    {code: "MD", name: "Moldova, Republic of"},
    {code: "MC", name: "Monaco"},
    {code: "MN", name: "Mongolia"},
    {code: "ME", name: "Montenegro"},
    {code: "MS", name: "Montserrat"},
    {code: "MA", name: "Morocco"},
    {code: "MZ", name: "Mozambique"},
    {code: "MM", name: "Myanmar"},
    {code: "NA", name: "Namibia"},
    {code: "NR", name: "Nauru"},
    {code: "NP", name: "Nepal"},
    {code: "NL", name: "Netherlands"},
    {code: "NC", name: "New Caledonia"},
    {code: "NZ", name: "New Zealand"},
    {code: "NI", name: "Nicaragua"},
    {code: "NE", name: "Niger"},
    {code: "NG", name: "Nigeria"},
    {code: "NU", name: "Niue"},
    {code: "NF", name: "Norfolk Island"},
    {code: "MP", name: "Northern Mariana Islands"},
    {code: "NO", name: "Norway"},
    {code: "OM", name: "Oman"},
    {code: "PK", name: "Pakistan"},
    {code: "PW", name: "Palau"},
    {code: "PS", name: "Palestinian Territory, Occupied"},
    {code: "PA", name: "Panama"},
    {code: "PG", name: "Papua New Guinea"},
    {code: "PY", name: "Paraguay"},
    {code: "PE", name: "Peru"},
    {code: "PH", name: "Philippines"},
    {code: "PN", name: "Pitcairn"},
    {code: "PL", name: "Poland"},
    {code: "PT", name: "Portugal"},
    {code: "PR", name: "Puerto Rico"},
    {code: "QA", name: "Qatar"},
    {code: "RE", name: "Réunion"},
    {code: "RO", name: "Romania"},
    {code: "RU", name: "Russian Federation"},
    {code: "RW", name: "Rwanda"},
    {code: "BL", name: "Saint Barthélemy"},
    {code: "SH", name: "Saint Helena, Ascension and Tristan da Cunha"},
    {code: "KN", name: "Saint Kitts and Nevis"},
    {code: "LC", name: "Saint Lucia"},
    {code: "MF", name: "Saint Martin (French part)"},
    {code: "PM", name: "Saint Pierre and Miquelon"},
    {code: "VC", name: "Saint Vincent and the Grenadines"},
    {code: "WS", name: "Samoa"},
    {code: "SM", name: "San Marino"},
    {code: "ST", name: "Sao Tome and Principe"},
    {code: "SA", name: "Saudi Arabia"},
    {code: "SN", name: "Senegal"},
    {code: "RS", name: "Serbia"},
    {code: "SC", name: "Seychelles"},
    {code: "SL", name: "Sierra Leone"},
    {code: "SG", name: "Singapore"},
    {code: "SX", name: "Sint Maarten (Dutch part)"},
    {code: "SK", name: "Slovakia"},
    {code: "SI", name: "Slovenia"},
    {code: "SB", name: "Solomon Islands"},
    {code: "SO", name: "Somalia"},
    {code: "ZA", name: "South Africa"},
    {code: "GS", name: "South Georgia and the South Sandwich Islands"},
    {code: "SS", name: "South Sudan"},
    {code: "ES", name: "Spain"},
    {code: "LK", name: "Sri Lanka"},
    {code: "SD", name: "Sudan"},
    {code: "SR", name: "Suriname"},
    {code: "SJ", name: "Svalbard and Jan Mayen"},
    {code: "SZ", name: "Swaziland"},
    {code: "SE", name: "Sweden"},
    {code: "CH", name: "Switzerland"},
    {code: "SY", name: "Syrian Arab Republic"},
    {code: "TW", name: "Taiwan, Province of China"},
    {code: "TJ", name: "Tajikistan"},
    {code: "TZ", name: "Tanzania, United Republic of"},
    {code: "TH", name: "Thailand"},
    {code: "TL", name: "Timor-Leste"},
    {code: "TG", name: "Togo"},
    {code: "TK", name: "Tokelau"},
    {code: "TO", name: "Tonga"},
    {code: "TT", name: "Trinidad and Tobago"},
    {code: "TN", name: "Tunisia"},
    {code: "TR", name: "Turkey"},
    {code: "TM", name: "Turkmenistan"},
    {code: "TC", name: "Turks and Caicos Islands"},
    {code: "TV", name: "Tuvalu"},
    {code: "UG", name: "Uganda"},
    {code: "UA", name: "Ukraine"},
    {code: "AE", name: "United Arab Emirates"},
    {code: "GB", name: "United Kingdom"},
    {code: "US", name: "United States"},
    {code: "UM", name: "United States Minor Outlying Islands"},
    {code: "UY", name: "Uruguay"},
    {code: "UZ", name: "Uzbekistan"},
    {code: "VU", name: "Vanuatu"},
    {code: "VE", name: "Venezuela, Bolivarian Republic of"},
    {code: "VN", name: "Viet Nam"},
    {code: "VG", name: "Virgin Islands, British"},
    {code: "VI", name: "Virgin Islands, U.S."},
    {code: "WF", name: "Wallis and Futuna"},
    {code: "EH", name: "Western Sahara"},
    {code: "YE", name: "Yemen"},
    {code: "ZM", name: "Zambia"},
    {code: "ZW", name: "Zimbabwe"}
];
